File: src/main/java/com/github/novotnyr/mavenversion/MavenVersionProjectViewDecorator.java
package com.github.novotnyr.mavenversion;

import com.intellij.ide.projectView.PresentationData;
import com.intellij.ide.projectView.ProjectViewNode;
import com.intellij.ide.projectView.ProjectViewNodeDecorator;
import com.intellij.openapi.util.SystemInfo;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.ui.SimpleTextAttributes;
import org.jetbrains.idea.maven.model.MavenId;
import org.jetbrains.idea.maven.project.MavenProject;
import org.jetbrains.idea.maven.project.MavenProjectsManager;

import java.util.Optional;
import java.util.function.Consumer;

public class MavenVersionProjectViewDecorator implements ProjectViewNodeDecorator {
    @Override
    public void decorate(ProjectViewNode<?> node, PresentationData data) {
        Optional.of(node)
                .filter(this::isDecorating)
                .map(ProjectViewNode::getProject)
                .map(MavenProjectsManager::getInstance)
                .filter(MavenProjectsManager::isMavenizedProject)
                .flatMap(mavenProjectsManager -> getPomXml(node)
                        .map(mavenProjectsManager::findProject)
                        .flatMap(MavenVersionProjectViewDecorator.this::getVersion))
                .ifPresent(doDecorate(data));
    }

    private boolean isDecorating(@SuppressWarnings("unused") ProjectViewNode<?> node) {
        return PluginSettings.getInstance().isShowVersion();
    }

    private Consumer<String> doDecorate(PresentationData data) {
        String separator = SystemInfo.isWindows ? "  " : "\t";
        return version -> data.addText(separator + version, SimpleTextAttributes.GRAY_ATTRIBUTES);
    }

    private Optional<String> getVersion(MavenProject mavenProject) {
        return Optional.of(mavenProject.getMavenId())
                .map(MavenId::getVersion);
    }

    private Optional<VirtualFile> getPomXml(ProjectViewNode<?> node) {
        return Optional.ofNullable(node)
                .map(ProjectViewNode::getVirtualFile)
                .map(virtualFile -> virtualFile.findChild("pom.xml"));
    }
}

File: src/main/kotlin/com/github/novotnyr/mavenversion/MavenVersionProjectViewDecorator.kt
package com.github.novotnyr.mavenversion

import com.intellij.ide.projectView.PresentationData
import com.intellij.ide.projectView.ProjectViewNode
import com.intellij.ide.projectView.ProjectViewNodeDecorator
import com.intellij.openapi.application.ApplicationInfo
import com.intellij.openapi.components.service
import com.intellij.openapi.util.SystemInfo
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.ui.SimpleTextAttributes
import org.jetbrains.idea.maven.project.MavenProjectsManager

private const val POM_XML = "pom.xml"

class MavenVersionProjectViewDecorator : ProjectViewNodeDecorator {

    override fun decorate(node: ProjectViewNode<*>, data: PresentationData) {
        if (!isDecorating()) return

        val project = node.project ?: return
        val mavenProjectsManager = MavenProjectsManager.getInstance(project)
        if (!mavenProjectsManager.isMavenizedProject) return

        node.pomXml
            ?.let { mavenProjectsManager.findProject(it) }
            ?.mavenId?.version
            ?.let { version ->
                data.addText("$separator$version", SimpleTextAttributes.GRAY_ATTRIBUTES)
            }
    }

    private fun isDecorating(): Boolean = service<PluginSettings>().state.showVersion

    private val ProjectViewNode<*>.pomXml: VirtualFile?
        get() = virtualFile?.findChild(POM_XML)

    private val separator: String
        get() = when {
            SystemInfo.isWindows -> "  "
            ApplicationInfo.getInstance().build.baselineVersion >= 261 -> " "
            else -> "\t"
        }
}